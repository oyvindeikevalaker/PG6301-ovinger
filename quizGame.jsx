import React, { useContext, useState } from "react";
import {
  BrowserRouter,
  Link,
  Route,
  Routes,
  useNavigate,
} from "react-router-dom";
import { isCorrectAnswer, randomQuestion } from "./questions";

export const QuestionContext = React.createContext({ randomQuestion });

export function FrontPage({ correctAnswers, questionsAnswered }) {
  return (
    <div>
      <h1>Quiz app</h1>
      <div data-testid={"status"}>
        You have answered {correctAnswers} of {questionsAnswered} correctly
      </div>
      <Link to={"/question"}>
        <button>Take a new quiz</button>
      </Link>
    </div>
  );
}

export function ShowQuestion({ setCorrectAnswers, setQuestionsAnswered }) {
  function handleAnswer(answer) {
    setQuestionsAnswered((q) => q + 1);
    if (isCorrectAnswer(question, answer)) {
      setCorrectAnswers((q) => q + 1);
      navigate("/answer/correct");
    } else {
      navigate("/answer/wrong");
    }
  }
  const navigate = useNavigate();
  const { randomQuestion } = useContext(QuestionContext);
  const [question] = useState(randomQuestion());
  return (
    <div>
      <h1>{question.question}</h1>
      {Object.keys(question.answers)
        .filter((a) => question.answers[a])
        .map((a) => (
          <div key={a} data-testid={a}>
            <button onClick={() => handleAnswer(a)}>
              {question.answers[a]}
            </button>
          </div>
        ))}
    </div>
  );
}

function ShowAnswer() {
  return (
    <div>
      <Routes>
        <Route path={"correct"} element={<h1>Correct!</h1>} />
        <Route path={"wrong"} element={<h1>Wrong!</h1>} />
      </Routes>
      <div>
        <Link to={"/"}>Show score</Link>
      </div>
      <div>
        <Link to={"/question"}>New question</Link>
      </div>
    </div>
  );
}

export function QuizGame() {
  const [questionsAnswered, setQuestionsAnswered] = useState(0);
  const [correctAnswers, setCorrectAnswers] = useState(0);
  return (
    <Routes>
      <Route
        path={"/"}
        element={
          <FrontPage
            questionsAnswered={questionsAnswered}
            correctAnswers={correctAnswers}
          />
        }
      />
      <Route
        path={"/question"}
        element={
          <ShowQuestion
            setQuestionsAnswered={setQuestionsAnswered}
            setCorrectAnswers={setCorrectAnswers}
          />
        }
      />
      <Route path={"/answer/*"} element={<ShowAnswer />} />
    </Routes>
  );
}
