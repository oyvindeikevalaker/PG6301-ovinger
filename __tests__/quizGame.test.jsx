import React from "react";
import ReactDOM from "react-dom";
import {FrontPage, QuestionContext, QuizGame} from "../quizGame";
import {MemoryRouter} from "react-router-dom";
import pretty from "pretty";

describe("Quiz game", () => {
    it("Shows answer status", () => {
        const element = document.createElement("div");
        ReactDOM.render(
            <MemoryRouter><FrontPage correctAnswers={3} questionsAnswered={10} /></MemoryRouter>,
            element
        );
        expect(element.querySelector("[data-testid=status]").textContent)
            .toEqual("You have answered 3 of 10 correctly");
        expect(pretty(element.innerHTML)).toMatchSnapshot();
    });

    it("shows question", () => {
        const question = {
            question: "Are you happy?",
            answers: {
                answer_a: "Yes",
                answer_b: "No",
                answer_c: "Maybe",
            }
        }
        const element = document.createElement("div");
        ReactDOM.render(
            <MemoryRouter initialEntries={["/question"]}>
                <QuestionContext.Provider value={{randomQuestion: () => question}}>
                    <QuizGame/>
                </QuestionContext.Provider>
            </MemoryRouter>,
            element
        );
        expect(pretty(element.innerHTML)).toMatchSnapshot();
    })
})