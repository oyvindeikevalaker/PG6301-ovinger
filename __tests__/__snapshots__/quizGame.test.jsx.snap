// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz game Shows answer status 1`] = `
"<div>
  <h1>Quiz app</h1>
  <div data-testid=\\"status\\">You have answered 3 of 10 correctly</div><a href=\\"/question\\"><button>Take a new quiz</button></a>
</div>"
`;

exports[`Quiz game shows question 1`] = `
"<div>
  <h1>Are you happy?</h1>
  <div><button>Yes</button></div>
  <div><button>No</button></div>
  <div><button>Maybe</button></div>
</div>"
`;
